// vue-auth-component/src/vue-auth-component.js
import AuthContainer from './auth-container/auth-container'

import LocalLoginForm from './forms/local-login/local-login'
import LocalSignupForm from './forms/local-signup/local-signup'

import AuthTabs from './tabs/tabs'

import AuthButton from './buttons/button'
import AmazonButton from './buttons/amazon/amazon'
import DropboxButton from './buttons/dropbox/dropbox'
import EvernoteButton from './buttons/evernote/evernote'
import FacebookButton from './buttons/facebook/facebook'
import GithubButton from './buttons/github/github'
import GoogleButton from './buttons/google/google'
import LinkedInButton from './buttons/linkedin/linkedin'
import MicrosoftButton from './buttons/microsoft/microsoft'
import OpenIDButton from './buttons/openid/openid'
import PayPalButton from './buttons/paypal/paypal'
import SkypeButton from './buttons/skype/skype'
import SlackButton from './buttons/slack/slack'
import StackOverflowButton from './buttons/stackoverflow/stackoverflow'
import TwitterButton from './buttons/twitter/twitter'
import YahooButton from './buttons/yahoo/yahoo'

export {
  AuthContainer,
  AuthTabs,
  LocalLoginForm,
  LocalSignupForm,
  AuthButton,
  AmazonButton,
  DropboxButton,
  EvernoteButton,
  FacebookButton,
  GithubButton,
  GoogleButton,
  LinkedInButton,
  MicrosoftButton,
  OpenIDButton,
  PayPalButton,
  SkypeButton,
  SlackButton,
  StackOverflowButton,
  TwitterButton,
  YahooButton
}

export default {
  AuthContainer,
  AuthTabs,
  LocalLoginForm,
  LocalSignupForm,
  AuthButton,
  AmazonButton,
  DropboxButton,
  EvernoteButton,
  FacebookButton,
  GithubButton,
  GoogleButton,
  LinkedInButton,
  MicrosoftButton,
  OpenIDButton,
  PayPalButton,
  SkypeButton,
  SlackButton,
  StackOverflowButton,
  TwitterButton,
  YahooButton
}
